src/solve_ac_powerflow.jl

```diff
@@ -0,0 +1,263 @@
+            data.bus_magnitude[:, time_step] .= NaN
+            data.bus_angles[:, time_step] .= NaN
+        end
+    end
+
+    # write branch flows
+    # TODO if Yft, Ytf change between time steps, this must be moved inside the loop!
+    ts_V =
+        data.bus_magnitude[:, sorted_time_steps] .*
+        exp.(1im .* data.bus_angles[:, sorted_time_steps])
+    Sft = ts_V[fb, :] .* conj.(Yft * ts_V)
+    Stf = ts_V[tb, :] .* conj.(Ytf * ts_V)
+
+    data.branch_activepower_flow_from_to .= real.(Sft)
+    data.branch_reactivepower_flow_from_to .= imag.(Sft)
+    data.branch_activepower_flow_to_from .= real.(Stf)
+    data.branch_reactivepower_flow_to_from .= imag.(Stf)
+
+    data.converged .= ts_converged
+
+    return
+end
+
+function _ac_powerflow(
+    data::ACPowerFlowData,
+    pf::ACPowerFlow{<:ACPowerFlowSolverType},
+    time_step::Int64;
+    kwargs...,
+)
+    check_reactive_power_limits = get(kwargs, :check_reactive_power_limits, false)
+
+    for _ in 1:MAX_REACTIVE_POWER_ITERATIONS
+        converged = _newton_powerflow(pf, data, time_step; kwargs...)
+        if !converged || !check_reactive_power_limits ||
+           _check_q_limit_bounds!(data, time_step)
+            return converged
+        end
+    end
+
+    @error("could not enforce reactive power limits after $MAX_REACTIVE_POWER_ITERATIONS")
+    return converged
+end
+
+function _check_q_limit_bounds!(
+    data::ACPowerFlowData,
+    time_step::Int64,
+)
+    bus_names = data.power_network_matrix.axes[1]
+    within_limits = true
+    bus_types = view(data.bus_type, :, time_step)
+    for (ix, bt) in enumerate(bus_types)
+        if bt == PSY.ACBusTypes.PV
+            Q_gen = data.bus_reactivepower_injection[ix, time_step]
+        else
+            continue
+        end
+
+        if Q_gen <= data.bus_reactivepower_bounds[ix, time_step][1]
+            @info "Bus $(bus_names[ix]) changed to PSY.ACBusTypes.PQ"
+            within_limits = false
+            data.bus_type[ix, time_step] = PSY.ACBusTypes.PQ
+            data.bus_reactivepower_injection[ix, time_step] =
+                data.bus_reactivepower_bounds[ix, time_step][1]
+        elseif Q_gen >= data.bus_reactivepower_bounds[ix, time_step][2]
+            @info "Bus $(bus_names[ix]) changed to PSY.ACBusTypes.PQ"
+            within_limits = false
+            data.bus_type[ix, time_step] = PSY.ACBusTypes.PQ
+            data.bus_reactivepower_injection[ix, time_step] =
+                data.bus_reactivepower_bounds[ix, time_step][2]
+        else
+            @debug "Within Limits"
+        end
+    end
+    return within_limits
+end
+
+function bus_type_idx(
+    data::ACPowerFlowData,
+    time_step::Int64 = 1,
+    bus_types::Tuple{Vararg{PSY.ACBusTypes}} = (
+        PSY.ACBusTypes.REF,
+        PSY.ACBusTypes.PV,
+        PSY.ACBusTypes.PQ,
+    ),
+)
+    # Find indices for each bus type
+    return [
+        findall(x -> x == bus_type, data.bus_type[:, time_step]) for bus_type in bus_types
```

Comment:
nitpick: cool functional programming thing you can do if you're interested:
```suggestion
        findall(==(bus_type), data.bus_type[:, time_step]) for bus_type in bus_types
```